src/Microsoft.ML.Transforms/WhiteningTransform.cs

```diff
@@ -207,139 +241,229 @@ private static VersionInfo GetVersionInfo()
                 loaderAssemblyName: typeof(WhiteningTransform).Assembly.FullName);
         }
 
-        private readonly ColInfoEx[] _exes;
-
-        private const string RegistrationName = "Whitening";
+        private readonly ColInfo[] _infos;
 
         /// <summary>
-        /// Convenience constructor for public facing API.
+        /// Initializes a new <see cref="WhiteningTransform"/> object.
         /// </summary>
         /// <param name="env">Host Environment.</param>
-        /// <param name="input">Input <see cref="IDataView"/>. This is the output from previous transform or loader.</param>
-        /// <param name="name">Name of the output column.</param>
-        /// <param name="source">Name of the column to be transformed. If this is null '<paramref name="name"/>' will be used.</param>
-        /// <param name="kind">Whitening kind (PCA/ZCA).</param>
-        public WhiteningTransform(IHostEnvironment env,
-            IDataView input,
-            string name,
-            string source = null,
-            WhiteningKind kind = Defaults.Kind)
-                : this(env, new Arguments() { Column = new[] { new Column() { Source = source ?? name, Name = name } }, Kind = kind }, input)
-        {
-        }
-
-        /// <summary>
-        /// Public constructor corresponding to SignatureDataTransform.
-        /// </summary>
-        public WhiteningTransform(IHostEnvironment env, Arguments args, IDataView input)
-            : base(env, RegistrationName, Contracts.CheckRef(args, nameof(args)).Column,
-                input, TestColumn)
+        /// <param name="inputData">Input data.</param>
+        /// <param name="columns">Describes the settings for the transformer.</param>
+        internal WhiteningTransform(IHostEnvironment env, IDataView inputData, params ColInfo[] columns)
```

Comment:
Why are take IDataView in the constructor of the WhiteningTransform? This should be needed later when calling Transform or create (i.e. when constructing the DataTransform).

Reply:
If you see NAReplaceTransform, it is also the case. The reason is that these are trainabile transforms, so it would not make sense to instantiate them without training them (what would you save? Or how would you transform?). Furthermore, this is just an internal constructor, the public facing API is consistent with estimators.

Reply:
So I would talk with Pete to make sure (in fact I am dealing with a similar issue with TermLookup). I believe the problem is that you are creating state in the Transform class. When you save the input schema and generate the Whitening data, you are creating state within the transform which you dont want to do as Transforms are supposed to be stateless. Its the DataTransform that will contain the state (and whitening data). 

If thats the case, you would need to move the whitening data generation code to either the Mapper (which I think would be the easiest, but may not make the most sense since a mapper is related to schema mapping), or implement your own DataTransform class that implements IRowToRowMapper which would allow you to customize the DataTransform to build the whitening data.

---
In reply to: [229552333](https://github.com/dotnet/machinelearning/pull/1452#discussion_r229552333) [](ancestors = 229552333)

Reply:
I think it is necessary for transforms to be stateful. A transform is the output of . fit (data) on an estimator. And that is the trained transform, which has to hold a state to further transform new data or to be saved and reloaded in another context.